File: telegram/src/main/java/by/mrrockka/service/FinalePlacesTelegramService.java
package by.mrrockka.service;

import by.mrrockka.domain.MessageMetadata;
import by.mrrockka.domain.TelegramPerson;
import by.mrrockka.domain.finaleplaces.FinalPlace;
import by.mrrockka.domain.finaleplaces.FinalePlaces;
import by.mrrockka.mapper.finaleplaces.FinalePlacesMessageMapper;
import by.mrrockka.response.builder.FinalePlacesResponseBuilder;
import by.mrrockka.service.exception.ChatGameNotFoundException;
import by.mrrockka.service.exception.FinalPlaceContainsNicknameOfNonExistingPlayerException;
import by.mrrockka.service.game.GameTelegramFacadeService;
import by.mrrockka.validation.GameValidator;
import by.mrrockka.validation.collection.CollectionsValidator;
import by.mrrockka.validation.mentions.PersonMentionsValidator;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.telegram.telegrambots.meta.api.methods.botapimethods.BotApiMethodMessage;
import org.telegram.telegrambots.meta.api.methods.send.SendMessage;

import java.util.List;

@Service
@RequiredArgsConstructor
public class FinalePlacesTelegramService {

  private final FinalePlacesService finalePlacesService;
  private final FinalePlacesMessageMapper finalePlacesMessageMapper;
  private final GameTelegramFacadeService gameTelegramFacadeService;
  private final TelegramPersonService telegramPersonService;
  private final FinalePlacesResponseBuilder finalePlacesResponseBuilder;
  private final GameValidator gameValidator;
  private final PersonMentionsValidator personMentionsValidator;
  private final CollectionsValidator collectionsValidator;

  public BotApiMethodMessage storePrizePool(final MessageMetadata messageMetadata) {
    personMentionsValidator.validateMessageMentions(messageMetadata, 1);

    final var places = finalePlacesMessageMapper.map(messageMetadata);
    collectionsValidator.validateMapIsNotEmpty(places, "Finale places");

    final var telegramGame = gameTelegramFacadeService
      .getGameByMessageMetadata(messageMetadata)
      .orElseThrow(ChatGameNotFoundException::new);
    gameValidator.validateGameIsTournamentType(telegramGame.game());

    final var telegramPersons = telegramPersonService
      .getAllByNicknamesAndChatId(places.values().stream().map(TelegramPerson::getNickname).toList(),
                                  messageMetadata.chatId());

    final var finalePlaces = new FinalePlaces(
      places.entrySet().stream()
        .map(place -> FinalPlace.builder()
          .position(place.getKey())
          .person(findByNickname(telegramPersons, place.getValue().getNickname()))
          .build())
        .toList());

    finalePlacesService.store(telegramGame.game().getId(), finalePlaces);
    return SendMessage.builder()
      .chatId(messageMetadata.chatId())
      .text(finalePlacesResponseBuilder.response(finalePlaces))
      .replyToMessageId(telegramGame.messageMetadata().id())
      .build();
  }

  private TelegramPerson findByNickname(final List<TelegramPerson> telegramPersons, final String nickname) {
    return telegramPersons.stream()
      .filter(person -> person.getNickname().equals(nickname))
      .findAny()
      .orElseThrow(() -> new FinalPlaceContainsNicknameOfNonExistingPlayerException(nickname));
  }

}

File: telegram/src/main/java/by/mrrockka/service/FinalePlacesTelegramService.kt
package by.mrrockka.service

import by.mrrockka.domain.CashGame
import by.mrrockka.domain.FinalPlace
import by.mrrockka.domain.MessageMetadata
import by.mrrockka.parser.FinalePlacesMessageParser
import by.mrrockka.repo.FinalePlacesRepo
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Propagation
import org.springframework.transaction.annotation.Transactional

interface FinalePlacesTelegramService {
    fun store(message: MessageMetadata): List<FinalPlace>
    fun store(message: MessageMetadata, finalePlaces: Map<Int, String>): List<FinalPlace>
}

@Service
@Transactional(propagation = Propagation.REQUIRED)
open class FinalePlacesTelegramServiceImpl(
        private val finalePlacesRepo: FinalePlacesRepo,
        private val finalePlacesParser: FinalePlacesMessageParser,
        private val gameService: GameTelegramService,
        private val personService: TelegramPersonService,
) : FinalePlacesTelegramService {

    override fun store(message: MessageMetadata): List<FinalPlace> {
        message.checkMentions()
        val places = finalePlacesParser.parse(message)
        return store(message, places)
    }

    override fun store(message: MessageMetadata, finalePlaces: Map<Int, String>): List<FinalPlace> {
        check(finalePlaces.isNotEmpty()) { "Finale places could not be empty" }

        val game = gameService.findGame(message)
        check(game !is CashGame) { "Finale places is not allowed for cash game" }
        val persons = personService.findByNicknames(finalePlaces.values.toList(), message.chatId)
                .associateBy { it.nickname }
        val mapped = finalePlaces
                .map { (position, nickname) ->
                    FinalPlace(position, persons[nickname] ?: error("Person not found for $nickname"))
                }
        finalePlacesRepo.store(game.id, mapped)

        return mapped
    }
}
